// hap-controller/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HomekitControllerError = exports.PairingTypeFlags = exports.PairMethods = exports.TLV = exports.Service = exports.IPDiscovery = exports.HttpConstants = exports.HttpClient = exports.GattUtils = exports.GattConstants = exports.GattClient = exports.Characteristic = exports.Category = exports.BLEDiscovery = void 0;
const ble_discovery_1 = __importDefault(require("./transport/ble/ble-discovery"));
exports.BLEDiscovery = ble_discovery_1.default;
const gatt_client_1 = __importDefault(require("./transport/ble/gatt-client"));
exports.GattClient = gatt_client_1.default;
const http_client_1 = __importDefault(require("./transport/ip/http-client"));
exports.HttpClient = http_client_1.default;
const pairing_protocol_1 = require("./protocol/pairing-protocol");
Object.defineProperty(exports, "PairingTypeFlags", { enumerable: true, get: function () { return pairing_protocol_1.PairingTypeFlags; } });
Object.defineProperty(exports, "PairMethods", { enumerable: true, get: function () { return pairing_protocol_1.PairMethods; } });
const ip_discovery_1 = __importDefault(require("./transport/ip/ip-discovery"));
exports.IPDiscovery = ip_discovery_1.default;
const Category = __importStar(require("./model/category"));
exports.Category = Category;
const Characteristic = __importStar(require("./model/characteristic"));
exports.Characteristic = Characteristic;
const GattConstants = __importStar(require("./transport/ble/gatt-constants"));
exports.GattConstants = GattConstants;
const GattUtils = __importStar(require("./transport/ble/gatt-utils"));
exports.GattUtils = GattUtils;
const HttpConstants = __importStar(require("./transport/ip/http-constants"));
exports.HttpConstants = HttpConstants;
const Service = __importStar(require("./model/service"));
exports.Service = Service;
const TLV = __importStar(require("./model/tlv"));
exports.TLV = TLV;
const error_1 = __importDefault(require("./model/error"));
exports.HomekitControllerError = error_1.default;
