// hap-controller/lib/model/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class HomekitControllerError extends Error {
    constructor(message, statusCode, body) {
        super(message);
        // eslint-disable-next-line no-undefined
        if (statusCode !== undefined) {
            this.setStatusCode(statusCode);
        }
        // eslint-disable-next-line no-undefined
        if (body !== undefined) {
            this.setBody(body);
        }
    }
    setStatusCode(errorCode) {
        this.statusCode = errorCode;
    }
    getStatusCode() {
        return this.statusCode;
    }
    setBody(body) {
        if (Buffer.isBuffer(body)) {
            try {
                this.body = JSON.parse(body.toString('utf-8'));
            }
            catch (err) {
                this.body = {
                    raw: body,
                };
            }
        }
        else {
            this.body = body;
        }
    }
    getBody() {
        return this.body;
    }
}
exports.default = HomekitControllerError;
