// hap-controller/lib/model/error.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const json_bigint_1 = __importDefault(require("json-bigint"));
class HomekitControllerError extends Error {
    constructor(message, statusCode, body) {
        super(message);
        // eslint-disable-next-line no-undefined
        if (statusCode !== undefined) {
            this.setStatusCode(statusCode);
        }
        // eslint-disable-next-line no-undefined
        if (body !== undefined) {
            this.setBody(body);
        }
    }
    setStatusCode(errorCode) {
        this.statusCode = errorCode;
    }
    getStatusCode() {
        return this.statusCode;
    }
    setBody(body) {
        if (Buffer.isBuffer(body)) {
            try {
                this.body = json_bigint_1.default.parse(body.toString('utf-8'));
            }
            catch (err) {
                this.body = {
                    raw: body,
                };
            }
        }
        else {
            this.body = body;
        }
    }
    getBody() {
        return this.body;
    }
}
exports.default = HomekitControllerError;
